pkg/controller/sync/controller.go

```diff
@@ -566,54 +533,98 @@ func (s *FederationSyncController) versionName(resourceName string) string {
 
-	selectedClusters, unselectedClusters := computePlacement(adapter.Placement(), placement, clusterNames)
+	selectedClusters, unselectedClusters := computePlacement(s.adapter.Placement(), placement,
+		clusterNames)
+
+	propagatedClusterVersions := getClusterVersions(s.adapter, template, override, propagatedVersion)
 
-	operations, err := operationsAccessor(adapter, selectedClusters, unselectedClusters, template, placement, override)
+	operations, err := s.clusterOperations(selectedClusters, unselectedClusters, template,
+		override, key, propagatedClusterVersions)
 	if err != nil {
+		s.eventRecorder.Eventf(template, corev1.EventTypeWarning, "FedClusterOperationsError",
+			"Error obtaining sync operations for %s: %s error: %s", kind, key, err.Error())
 		return statusError
 	}
 
 	if len(operations) == 0 {
 		return statusAllOK
 	}
 
-	clusterVersions, operationErrors := execute(operations)
+	updatedClusterVersions, operationErrors := s.updater.Update(operations)
 
-	err = updatePropagatedVersion(adapter, clusterVersions, template, override, version, selectedClusters, pendingVersionUpdates)
+	err = updatePropagatedVersion(s.adapter, updatedClusterVersions, template, override,
+		propagatedVersion, selectedClusters, s.pendingVersionUpdates)
 	if err != nil {
-		runtime.HandleError(fmt.Errorf("Failed to record propagated version for %s %q: %v", kind, key, err))
+		runtime.HandleError(fmt.Errorf("Failed to record propagated version for %s %q: %v", kind,
+			key, err))
 		// Don't return an error - failure to record the propagated
 		// version does not imply that propagation for the resource
 		// needs to be attempted again.
 	}
 
 	if len(operationErrors) > 0 {
-		runtime.HandleError(fmt.Errorf("Failed to execute updates for %s %q: %v", kind, key, operationErrors))
+		runtime.HandleError(fmt.Errorf("Failed to execute updates for %s %q: %v", kind,
+			key, operationErrors))
 		return statusError
 	}
 
 	return statusAllOK
 }
 
-func updatePropagatedVersion(adapter federatedtypes.FederatedTypeAdapter, versions map[string]string, template, override pkgruntime.Object, version *fedv1a1.PropagatedVersion, selectedClusters []string, pendingVersionUpdates sets.String) error {
+// getClusterVersions returns the cluster versions populated in the current
+// propagated version object if it exists.
+func getClusterVersions(adapter federatedtypes.FederatedTypeAdapter, template,
+	override pkgruntime.Object, propagatedVersion *fedv1a1.PropagatedVersion) map[string]string {
+
+	clusterVersions := make(map[string]string)
+
+	if propagatedVersion == nil {
+		return clusterVersions
+	}
+
+	templateVersion := adapter.Template().ObjectMeta(template).ResourceVersion
+	overrideVersion := ""
+
+	if override != nil {
+		overrideVersion = adapter.Override().ObjectMeta(override).ResourceVersion
+	}
+
+	if templateVersion == propagatedVersion.Status.TemplateVersion &&
+		overrideVersion == propagatedVersion.Status.OverrideVersion {
```

Comment:
(No action required) Ugh, I really don't like seeing the conditional line up with the contents.  I'd like to see a clear visual differentiation between the inside and outside of a block.

Reply:
Agreed. Perhaps an extra space would help as I've done to functions to split between the function parameters and its contents since opening braces are required to be on the same line. But again, trying to avoid these really long lines.

Personally, I prefer the boolean operator to indicate the beginning of continuing conditional expressions but it seems `gofmt` via my plugin does not like it.